src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -0,0 +1,460 @@
+        assert(mapped.RowPitch >= _r.cells.size() * sizeof(Cell));
+        memcpy(mapped.pData, _r.cells.data(), _r.cells.size() * sizeof(Cell));
+        _r.deviceContext->Unmap(_r.cellBuffer.get(), 0);
+    }
+
+    // After Present calls, the back buffer needs to explicitly be
+    // re-bound to the D3D11 immediate context before it can be used again.
+    _r.deviceContext->OMSetRenderTargets(1, _r.renderTargetView.addressof(), nullptr);
+    _r.deviceContext->Draw(3, 0);
+
+    // See documentation for IDXGISwapChain2::GetFrameLatencyWaitableObject method:
+    // > For every frame it renders, the app should wait on this handle before starting any rendering operations.
+    // > Note that this requirement includes the first frame the app renders with the swap chain.
+    assert(_r.frameLatencyWaitableObjectUsed);
+
+    // > IDXGISwapChain::Present: Partial Presentation (using a dirty rects or scroll) is not supported
+    // > for SwapChains created with DXGI_SWAP_EFFECT_DISCARD or DXGI_SWAP_EFFECT_FLIP_DISCARD.
+    // ---> No need to call IDXGISwapChain1::Present1.
+    //      TODO: Would IDXGISwapChain1::Present1 and its dirty rects have benefits for remote desktop?
+    THROW_IF_FAILED(_r.swapChain->Present(1, 0));
+
+    // On some GPUs with tile based deferred rendering (TBDR) architectures, binding
+    // RenderTargets that already have contents in them (from previous rendering) incurs a
+    // cost for having to copy the RenderTarget contents back into tile memory for rendering.
+    //
+    // On Windows 10 with DXGI_SWAP_EFFECT_FLIP_DISCARD we get this for free.
+    if (!_sr.isWindows10OrGreater)
+    {
+        _r.deviceContext->DiscardView(_r.renderTargetView.get());
+    }
+
+    return S_OK;
+}
+catch (const wil::ResultException& exception)
+{
+    return _handleException(exception);
+}
+CATCH_RETURN()
+
+#pragma endregion
+
+void AtlasEngine::_setShaderResources() const
+{
+    _r.deviceContext->VSSetShader(_r.vertexShader.get(), nullptr, 0);
+    _r.deviceContext->PSSetShader(_r.pixelShader.get(), nullptr, 0);
+
+    // Our vertex shader uses a trick from Bill Bilodeau published in
+    // "Vertex Shader Tricks" at GDC14 to draw a fullscreen triangle
+    // without vertex/index buffers. This prepares our context for this.
+    _r.deviceContext->IASetVertexBuffers(0, 0, nullptr, nullptr, nullptr);
+    _r.deviceContext->IASetIndexBuffer(nullptr, DXGI_FORMAT_UNKNOWN, 0);
+    _r.deviceContext->IASetInputLayout(nullptr);
+    _r.deviceContext->IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+
+    _r.deviceContext->PSSetConstantBuffers(0, 1, _r.constantBuffer.addressof());
+
+    const std::array resources{ _r.cellView.get(), _r.atlasView.get() };
+    _r.deviceContext->PSSetShaderResources(0, gsl::narrow_cast<UINT>(resources.size()), resources.data());
+}
+
+AtlasEngine::f32x4 AtlasEngine::_getGammaRatios(float gamma) noexcept
+{
+    static constexpr f32x4 gammaIncorrectTargetRatios[13]{

```

Comment:
>gammaIncorrectTargetRatios [](http://example.com/codeflow?start=27&length=26)

"Incorrect"?

Reply:
Yeah there's also gamma-correct correction, uhm, somewhere iykwim.
Now you don't start assuming that I know what I'm doing here, alright? I think - and that's a huge, the way Donald calls stuff huge - stretch of my saying "I think"... I think gamma incorrect correction applies to display-native gamma surfaces and gamma correct correction (lol) applies to sRGB surfaces. `gammaIncorrectTargetRatios` was """definitely""" the table to be used for D3D non-sRGB surfaces. I think. Maybe, but probably.

Reply:
Okay. I thought you were making a joke on the opposite of "correct" as you explained to me the inverse correction required over chat the other day.